### ISSUE TITLE
Cannot log in automatically on machine where Tomcat service is running

### ISSUE THREAD

**Issue by thisusernameisfree** [dblock/waffle#268]
I'm trying to set up windows authentication with waffle, and it works fine as long as I use LOCAL SYSTEM and NETWORK SERVICE account. But if I switch to the domain account I'm getting "The handle specified is invalid" error. 

I tried to configure SPNs on AD server but it doesn't help. I also enabled Kerberos delegation for that server, and for that account.
C:\Users\Administrator>setspn -l dbadmin
Registered ServicePrincipalNames for CN=dbadmin,CN=Users,DC=blah,DC=com:
        HTTP/Unicorn
        HTTP/Unicorn.blah.com

Here is the log with errors:
```
2015-10-19 19:13:14,825 DEBUG [http-nio-8085-exec-4] [waffle.servlet.NegotiateSecurityFilter] [::] -- GET /idp/login, contentlength: -1
2015-10-19 19:13:14,825 DEBUG [http-nio-8085-exec-4] [waffle.servlet.NegotiateSecurityFilter] [::] -- authorization required
2015-10-19 19:13:14,856 DEBUG [http-nio-8085-exec-5] [waffle.servlet.NegotiateSecurityFilter] [::] -- GET /idp/login, contentlength: -1
2015-10-19 19:13:14,856 DEBUG [http-nio-8085-exec-5] [waffle.servlet.spi.NegotiateSecurityFilterProvider] [::] -- security package: Negotiate, connection id: fe80:0:0:0:852:798e:74de:9d1e%12:50469
2015-10-19 19:13:14,856 DEBUG [http-nio-8085-exec-5] [waffle.servlet.spi.NegotiateSecurityFilterProvider] [::] -- token buffer: 119 byte(s)
2015-10-19 19:13:14,859 DEBUG [http-nio-8085-exec-5] [waffle.servlet.spi.NegotiateSecurityFilterProvider] [::] -- continue token: oYHWMIHToAMKAQGhDAYKKwYBBAGCNwICCqKBvQSBuk5UTE1TU1AAAgAAAAgACAA4AAAAFcKJ4ngIGibaBiMXILPmHCYAAAB6AHoAQAAAAAYDgCUAAAAPQgBMAEEASAACAAgAQgBMAEEASAABAA4AVQBOAEkAQwBPAFIATgAEABAAYgBsAGEAaAAuAGMAbwBtAAMAIABVAG4AaQBjAG8AcgBuAC4AYgBsAGEAaAAuAGMAbwBtAAUAEABiAGwAYQBoAC4AYwBvAG0ABwAIAE/a+g6JCtEBAAAAAA==
2015-10-19 19:13:14,859 DEBUG [http-nio-8085-exec-5] [waffle.servlet.spi.NegotiateSecurityFilterProvider] [::] -- continue required: true
2015-10-19 19:13:14,862 DEBUG [http-nio-8085-exec-6] [waffle.servlet.NegotiateSecurityFilter] [::] -- GET /idp/login, contentlength: -1
2015-10-19 19:13:14,862 DEBUG [http-nio-8085-exec-6] [waffle.servlet.spi.NegotiateSecurityFilterProvider] [::] -- security package: Negotiate, connection id: fe80:0:0:0:852:798e:74de:9d1e%12:50469
2015-10-19 19:13:14,862 DEBUG [http-nio-8085-exec-6] [waffle.servlet.spi.NegotiateSecurityFilterProvider] [::] -- token buffer: 121 byte(s)
2015-10-19 19:13:14,871 WARN  [http-nio-8085-exec-6] [waffle.servlet.NegotiateSecurityFilter] [::] -- error logging in user: com.sun.jna.platform.win32.Win32Exception: The handle specified is invalid
```
I think I'm missing something but I can't figure out what, is there any other privileges I need to give that service account? I made it as local administrator to eliminate possible issues on local machine where the service is running.

As a result I'm getting a popup window prompting to enter my credentials. When I enter user and password - I'm successfully logged in. I also don't experience any problems If I remove Negotiate from protocols and keep only NTLM.

**Comment by thisusernameisfree** [comment]
Here is a more detailed log with error I'm getting:
```
2015-10-20 14:15:54,431 DEBUG [http-nio-8085-exec-4] [waffle.servlet.NegotiateSecurityFilter] [::] -- GET /idp/login, contentlength: -1
2015-10-20 14:15:55,572 DEBUG [http-nio-8085-exec-4] [waffle.servlet.NegotiateSecurityFilter] [::] -- authorization required
2015-10-20 14:15:55,806 DEBUG [http-nio-8085-exec-5] [waffle.servlet.NegotiateSecurityFilter] [::] -- GET /idp/login, contentlength: -1
2015-10-20 14:15:55,830 DEBUG [http-nio-8085-exec-5] [waffle.servlet.spi.NegotiateSecurityFilterProvider] [::] -- security package: Negotiate, connection id: 10.136.41.130:50091
2015-10-20 14:15:55,856 DEBUG [http-nio-8085-exec-5] [waffle.servlet.spi.NegotiateSecurityFilterProvider] [::] -- token buffer: 119 byte(s)
2015-10-20 14:15:56,339 DEBUG [http-nio-8085-exec-5] [waffle.servlet.spi.NegotiateSecurityFilterProvider] [::] -- continue token: oYHWMIHToAMKAQGhDAYKKwYBBAGCNwICCqKBvQSBuk5UTE1TU1AAAgAAAAgACAA4AAAAFcKJ4kL4Rn8V+6bSkE9bhN0AAAB6AHoAQAAAAAYDgCUAAAAPQgBMAEEASAACAAgAQgBMAEEASAABAA4AVQBOAEkAQwBPAFIATgAEABAAYgBsAGEAaAAuAGMAbwBtAAMAIABVAG4AaQBjAG8AcgBuAC4AYgBsAGEAaAAuAGMAbwBtAAUAEABiAGwAYQBoAC4AYwBvAG0ABwAIAPmzybAoC9EBAAAAAA==
2015-10-20 14:15:56,340 DEBUG [http-nio-8085-exec-5] [waffle.servlet.spi.NegotiateSecurityFilterProvider] [::] -- continue required: true
2015-10-20 14:15:56,582 DEBUG [http-nio-8085-exec-6] [waffle.servlet.NegotiateSecurityFilter] [::] -- GET /idp/login, contentlength: -1
2015-10-20 14:15:56,582 DEBUG [http-nio-8085-exec-6] [waffle.servlet.spi.NegotiateSecurityFilterProvider] [::] -- security package: Negotiate, connection id: 10.136.41.130:50091
2015-10-20 14:15:56,588 DEBUG [http-nio-8085-exec-6] [waffle.servlet.spi.NegotiateSecurityFilterProvider] [::] -- token buffer: 121 byte(s)
2015-10-20 14:15:56,938 WARN  [http-nio-8085-exec-6] [waffle.servlet.NegotiateSecurityFilter] [::] -- error logging in user: com.sun.jna.platform.win32.Win32Exception: The handle specified is invalid
2015-10-20 14:15:56,939 TRACE [http-nio-8085-exec-6] [waffle.servlet.NegotiateSecurityFilter] [::] -- {}
java.io.IOException: com.sun.jna.platform.win32.Win32Exception: The handle specified is invalid
	at waffle.servlet.spi.SecurityFilterProviderCollection.doFilter(SecurityFilterProviderCollection.java:150)
	at waffle.servlet.NegotiateSecurityFilter.doFilter(NegotiateSecurityFilter.java:123)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.eclipse.alf.security.sso.server.federationserver.filter.LoginRedirectFilter.doFilter(LoginRedirectFilter.java:164)
	at org.apache.catalina.core.ApplicationFilterChain.internalDoFilter(ApplicationFilterChain.java:241)
	at org.apache.catalina.core.ApplicationFilterChain.doFilter(ApplicationFilterChain.java:208)
	at org.apache.catalina.core.StandardWrapperValve.invoke(StandardWrapperValve.java:220)
	at org.apache.catalina.core.StandardContextValve.invoke(StandardContextValve.java:122)
	at org.apache.catalina.core.StandardHostValve.invoke(StandardHostValve.java:170)
	at org.apache.catalina.valves.ErrorReportValve.invoke(ErrorReportValve.java:103)
	at org.apache.catalina.core.StandardEngineValve.invoke(StandardEngineValve.java:116)
	at org.apache.catalina.connector.CoyoteAdapter.service(CoyoteAdapter.java:423)
	at org.apache.coyote.http11.AbstractHttp11Processor.process(AbstractHttp11Processor.java:1079)
	at org.apache.coyote.AbstractProtocol$AbstractConnectionHandler.process(AbstractProtocol.java:625)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.doRun(NioEndpoint.java:1760)
	at org.apache.tomcat.util.net.NioEndpoint$SocketProcessor.run(NioEndpoint.java:1719)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at org.apache.tomcat.util.threads.TaskThread$WrappingRunnable.run(TaskThread.java:61)
	at java.lang.Thread.run(Thread.java:745)
Caused by: com.sun.jna.platform.win32.Win32Exception: The handle specified is invalid
	at waffle.windows.auth.impl.WindowsAuthProviderImpl.acceptSecurityToken(WindowsAuthProviderImpl.java:140)
	at waffle.servlet.spi.NegotiateSecurityFilterProvider.doFilter(NegotiateSecurityFilterProvider.java:144)
	at waffle.servlet.spi.SecurityFilterProviderCollection.doFilter(SecurityFilterProviderCollection.java:148)
	... 20 more
```

I have also enabled Kerberos logging but I don't see any errors in event viewer both on AD and local server. It looks like the token I supply cannot be accepted for some reason. Could it still be an issue with SPNs?

**Comment by dblock** [comment]
Yes this could be an SPN issue or a permissions problem for the service account, etc.

**Comment by thisusernameisfree** [comment]
I did some more testing and I found that I can login successfully if I try to do this from another machine in domain. However I can't login with this user on the machine where the service is running.

**Comment by dblock** [comment]
It's pretty clear that there's a local policy that is explicitly denying this. It would be great if you could find out which one.

**Comment by thisusernameisfree** [comment]
I'm still looking through the possible local policy restrictions, but here is some info about what happens when I send the request locally:

1) Client is requesting the web page GET http://Unicorn.blah.com/webpage.
2) Server returns 401 response with WWW-Authenticate headers: Negotiate, NTLM, Basic Realm
3) Client sends Authorization header:
```
Authorization Header (Negotiate) appears to contain a Kerberos ticket:
60 75 06 06 2B 06 01 05 05 02 A0 6B 30 69 A0 30  `u..+..... k0i 0
30 2E 06 0A 2B 06 01 04 01 82 37 02 02 0A 06 09  0...+....7.....
2A 86 48 82 F7 12 01 02 02 06 09 2A 86 48 86 F7  *H÷......*H÷
12 01 02 02 06 0A 2B 06 01 04 01 82 37 02 02 1E  ......+....7...
A2 35 04 33 4E 54 4C 4D 53 53 50 00 01 00 00 00  ¢5.3NTLMSSP.....
97 B2 08 E2 04 00 04 00 2F 00 00 00 07 00 07 00  ².â..../.......
28 00 00 00 06 03 80 25 00 00 00 0F 55 4E 49 43  (.....%....UNIC
4F 52 4E 42 4C 41 48                             ORNBLAH 
```
4) Server returns 401 with WWW-Authenticate header, probably prompting to continue:
```
WWW-Authenticate Header (Negotiate) appears to be a Kerberos reply:
A1 81 D6 30 81 D3 A0 03 0A 01 01 A1 0C 06 0A 2B  ¡Ö0Ó ....¡...+
06 01 04 01 82 37 02 02 0A A2 81 BD 04 81 BA 4E  ....7...¢½.ºN
54 4C 4D 53 53 50 00 02 00 00 00 08 00 08 00 38  TLMSSP.........8
00 00 00 15 C2 89 E2 A0 34 58 8C E2 7D D2 43 80  ....Ââ 4Xâ}ÒC
4F 56 53 A6 00 00 00 7A 00 7A 00 40 00 00 00 06  OVS¦...z.z.@....
03 80 25 00 00 00 0F 42 00 4C 00 41 00 48 00 02  .%....B.L.A.H..
00 08 00 42 00 4C 00 41 00 48 00 01 00 0E 00 55  ...B.L.A.H.....U
00 4E 00 49 00 43 00 4F 00 52 00 4E 00 04 00 10  .N.I.C.O.R.N....
00 62 00 6C 00 61 00 68 00 2E 00 63 00 6F 00 6D  .b.l.a.h...c.o.m
00 03 00 20 00 55 00 6E 00 69 00 63 00 6F 00 72  ... .U.n.i.c.o.r
00 6E 00 2E 00 62 00 6C 00 61 00 68 00 2E 00 63  .n...b.l.a.h...c
00 6F 00 6D 00 05 00 10 00 62 00 6C 00 61 00 68  .o.m.....b.l.a.h
00 2E 00 63 00 6F 00 6D 00 07 00 08 00 46 70 20  ...c.o.m.....Fp 
AA E3 0C D1 01 00 00 00 00                       ªã.Ñ.....       
```
5) Client sends another request with Authorization header:
```
Authorization Header (Negotiate) appears to contain a Kerberos ticket:
A1 77 30 75 A0 03 0A 01 01 A2 5A 04 58 4E 54 4C  ¡w0u ....¢Z.XNTL
4D 53 53 50 00 03 00 00 00 00 00 00 00 58 00 00  MSSP.........X..
00 00 00 00 00 58 00 00 00 00 00 00 00 58 00 00  .....X.......X..
00 00 00 00 00 58 00 00 00 00 00 00 00 58 00 00  .....X.......X..
00 00 00 00 00 58 00 00 00 15 C2 88 E2 06 03 80  .....X....Ââ..
25 00 00 00 0F EA 39 8C 92 1D B2 80 F2 FD 9E B2  %....ê9.²òý²
72 55 4A 3E 3E A3 12 04 10 01 00 00 00 F5 17 A7  rUJ>>£.......õ.§
50 2D 22 9A 84 00 00 00 00                       P-".... 
```
6) Server returns 401 response with WWW-Authenticate headers: Negotiate, NTLM, Basic Realm. And here it stops and prompts credentials.

I don't know what those tickets mean because I can't decipher them. However when I log in from another machine with the same user, the server returns 200 on step 4 (no need to continue). Also the first authorization header is much bigger and contains a lot of info, probably user and group details with whatever is needed for Kerberos.

Could local policies actually make a difference on what client sends to server?

**Comment by dblock** [comment]
Yes, when you're going from a machine to the same machine via HTTP you're still doing a network logon, but with special provisions (same machine to the same machine), so even local machine policy can affect what's going on. I just don't know how to find which specific privilege denies logon, but there's definitely one, unless it's an SPN issue.

**Comment by thisusernameisfree** [comment]
Researching it further I found that it works fine if I start Tomcat in console mode on behalf of the same domain user. It is reproducible on Windows Server 2008 and 2012 but not on 2003. It could be related to some new security features like session 0 isolation or something similar.

**Comment by thisusernameisfree** [comment]
This issue was resolved in the final update of the answer here:
http://stackoverflow.com/questions/33525493/acceptsecuritycontext-fails-when-application-is-running-as-a-service

**Comment by dblock** [comment]
@thisusernameisfree What does this mean for Waffle? Do we need any code changes? Can you contribute an FAQ please?

**Comment by thisusernameisfree** [comment]
Yes, this requires a code change that I mentioned above in WindowsAuthProviderImpl.java. The server credentials need to be acquired only once and then reused for all subsequent requests from the same client.
Here is my pull request for FAQ: https://github.com/dblock/waffle/pull/296

**Comment by dblock** [comment]
So this is really a bug, right? Why don't we actually fix it instead of making an FAQ? Want to PR it? Do all tests pass with extracting it?

**Comment by thisusernameisfree** [comment]
Yes, this is a bug and it is better to fix it. The change should be pretty simple so I don't really need to PR it. We can discard the FAQ merge.

**Comment by dblock** [comment]
Please contribute a PR with a CHANGELOG entry and such. The Waffle "team" is a bunch of volunteers, so we appreciate all the help we get.

**Comment by hazendaz** [comment]
@thisusernameisfree A fix was provided by @vimil has been applied to master.  Thanks for all the links and thorough background on this issue.

**Comment by hazendaz** [comment]
This was patched on 405943bad41b8d6b61d9d16728c42280c7a57aed
